### PR TITLE
File source indexer

### PR DESCRIPTION
**Pull request by adammcclenaghan**

### CODE REVIEW

File: syft/internal/fileresolver/file.go [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+				return nil, fmt.Errorf("unable to get file metadata for reference %s: %w", refVia.Reference.RealPath, err)
+			}
+
+			// don't consider directories
+			if entry.Metadata.IsDir() {
+				continue
+			}
+
+			loc := file.NewVirtualLocationFromDirectory(
+				r.responsePath(string(refVia.Reference.RealPath)), // the actual path relative to the resolver root
```
Comment (HairyMike) [inline syft/internal/fileresolver/file.go:178]: RealPath looks like a string - does it need to be cast?

File: syft/internal/fileresolver/file.go [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+// path is the filepath of the file we're creating content access for
+func NewFromFile(parent, path string, pathFilters ...PathIndexVisitor) (*File, error) {
+	chroot, err := NewChrootContextFromCWD(parent, parent)
+	if err != nil {
+		return nil, fmt.Errorf("unable to interpret chroot context: %w", err)
+	}
+
+	cleanBase := chroot.Base()
+
+	chroot.Base()
```
Comment (HairyMike) [inline syft/internal/fileresolver/file.go]: called twice?

File: syft/internal/fileresolver/file.go [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+		}
+
+		// don't consider directories
+		if entry.Metadata.IsDir() {
+			continue
+		}
+
+		if windows.HostRunningOnWindows() {
+			userStrPath = windows.ToPosix(userStrPath)
+		}
```
Comment (HairyMike) [inline syft/internal/fileresolver/file.go:127]: maybe want to abstract this method out so you don't have to repeat, e.g:
```
func normalizePathForOS(path string) string {
    if windows.HostRunningOnWindows() {
        return windows.ToPosix(path)
    }
    return path
}
```

File: syft/internal/fileresolver/file.go [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+}
+
+func (r File) requestGlob(pattern string) (string, error) {
+	return r.chroot.ToNativeGlob(pattern)
+}
+
+// FilesByGlob returns all file.References that match the given path glob pattern from any layer in the image.
+func (r File) FilesByGlob(patterns ...string) ([]file.Location, error) {
+	uniqueFileIDs := stereoscopeFile.NewFileReferenceSet()
+	uniqueLocations := make([]file.Location, 0)
```
Comment (HairyMike) [inline syft/internal/fileresolver/file.go:148]: I think you can preallocate this slice:
```
uniqueLocations := make([]file.Location, len(patterns))
```

File: syft/internal/fileresolver/file.go [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+	requestPath, err := r.requestPath(userPath)
+	if err != nil {
+		return false
+	}
+	return r.tree.HasPath(stereoscopeFile.Path(requestPath))
+}
+
+// Stringer to represent a file path data source
+func (r File) String() string {
+	return fmt.Sprintf("file:%s", r.path)
```
Comment (HairyMike) [inline syft/internal/fileresolver/file.go:89]: If this is a high frequency call, you could consider using a `strings.Builder` here as it needs fewer allocations.

**Comment by HairyMike** [comment]
Really just nitpick comments

File: syft/internal/fileresolver/file.go (hunk 1 above)
Comment (adammcclenaghan) [reply syft/internal/fileresolver/file.go:178]: Yeah, golang enforces it

File: syft/internal/fileresolver/file.go (hunk 3 above)
Comment (adammcclenaghan) [reply syft/internal/fileresolver/file.go:127]: Subtle difference - one of these we convert ToPosix, in the other its FromPosix

File: syft/internal/fileresolver/file.go (hunk 5 above)
Comment (adammcclenaghan) [reply syft/internal/fileresolver/file.go:89]: Just going to stick with this for now as it matches directory.go impl & I want to keep things as similar as possible to reduce congnitive overhead for maintainers reviewing - I'm also not sure about how frequently this is called so don't want to optimise prematurely

**Comment by adammcclenaghan** [comment]
Closing, will open PR to upstream now.

**Pull request closed without merging**